### DISCUSSION TITLE
RFC: Thoughts on databases.

### DISCUSSION THREAD

**Discussion by marciocadev** [winglang/wing#3313] [Ideas]
### Feature Spec

I was thinking about what @ekeren said about DB, and I agree with what he said.
 [the wise words of Eyal](https://github.com/winglang/wing/issues/2437#issuecomment-1581950309)

At first, using abstractions to work in the cloud is exciting, but when it comes to databases, trying to "improve" access with new abstractions tends to make it more difficult to use. I believe that the typical cloud resource user already has prior knowledge of using databases, and our current resource (Table) seems counterintuitive to me.

If you have worked with relational databases, the Table object alone doesn't mean much. If you have worked with non-relational databases like DynamoDB, our current implementation of the Table is counterintuitive (the primary key brings to mind associations with relational databases, not the partition key of DynamoDB).

I believe that the best way to create persistence resources for the language is to focus on abstractions such as permissioning and reflect the current implementations of each database.

When creating a relational database like PostgreSQL or MySQL, allow the creation of schemas, not individual tables.

Meanwhile, for non-relational databases, reflect the particularities of each one.

### Use Cases

Make the use of DB in Wing simpler.

### Component

SDK

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

**Comment by Joshswooft** [comment]
I'm also against creating an abstraction around the various databases. I think if you go down that route people will lose either the easiness of interacting with the DB through Wing or the power that comes from the unique aspects of their chosen DB.

**Comment by ShaiBer** [comment]
I'm also against abstracting DBs

**Comment by staycoolcall911** [comment]
Turning this into a discussion until we figure out some action items

**Comment by whatnick** [comment]
The term database is a bit too generic to be backed by a concrete abstraction. Perhaps bucketing the available classes of databases across cloud providers can be the source of a more actionable generic.
- Relational Databases - PostgreSQL, MySQL etc.
- Wide Column Table Oriented Databases - DynamoDB, BigTable etc.
- etc.
- The aim would be to also provide a concrete API downstream for applications using this abstraction to implement against. We are perhaps talking about something along the lines of a DBAPI model other languages have. Such as PEP 249 : https://peps.python.org/pep-0249/
